src/vis/bar/utils.ts

```diff
@@ -172,11 +175,20 @@ async function setPlotsWithGroupsAndMultiples(
 
       const plotAggregateAxisName = createAxisLabel(aggregateType, aggregateColumn);
 
+      let valIdArr = uniqueColVals.map((val) => []);
+
+      catColValues.resolvedValues.forEach((row) => valIdArr[uniqueColVals.indexOf(row.val as string)].push(row.id));
+
+      valIdArr = valIdArr.map((arr) => arr.filter((val) => allGroupObjsIds.has(val) && allMultiplesObjsIds.has(val)));
```

Comment:
Yes, totally agree with you the data wrangling logic in this file is out of hand. And also happens with very minor changes like four times. I've mentioned it before, but I'm not really sure what to do about it. I've never seen these libraries before, but they do look helpful! Definitely seems worth a look.

Reply:
We should try to find some time for a refactoring. Because even now, with some time in between, the data wrangling code is hard to follow and read between all the loops. We can also add some unit tests at least for the utils and/or break the large functions up in smaller pieces to make it testable/readable/maintainable.